### PR TITLE
Run algorithms inside Docker

### PR DESCRIPTION
**Pull request by erikbern**
not quite sure how this will evolve... will poke around until it works

### CODE REVIEW

**Comment by erikbern** [comment]
Shockingly, Travis built this one!

All it does so far is installing each library (and only Annoy and RPForest). Next step is to figure out how to run the benchmarks inside Docker (rather than the clunky subprocess way)

**Comment by erikbern** [comment]
rough outline of the rest:
1. rewrite `run_algo` so that it writes the results in the child process not the parent
2. `get_algorithms` should return references, not objects
3. `run_single.py` construct objects on the fly and run `run_algo`
4. rewrite `main` to use docker instead of multiprocessing

**Comment by erikbern** [comment]
All the algorithms are now building fine in Docker

What's missing is
1. Run algos inside Docker instead of subprocess
2. Run algos in Travis to make sure they also run well, not just install

**Comment by maumueller** [comment]
Do you know if there is a better way of getting the index size than https://github.com/erikbern/ann-benchmarks/blob/master/ann_benchmarks/algorithms/base.py#L16 when using docker?

**Comment by erikbern** [comment]
> Do you know if there is a better way of getting the index size than https://github.com/erikbern/ann-benchmarks/blob/master/ann_benchmarks/algorithms/base.py#L16 when using docker?

I think you can measure memory usage somehow through the Docker SDK: https://docker-py.readthedocs.io/en/stable/containers.html – not sure how reliable it is (memory usage is notoriously hard to measure)

**Comment by maumueller** [comment]
At least we could get some measurements for comparison with the method that is currently included.

**Comment by erikbern** [comment]
Btw it's so much nicer already using Docker. I'm able to install all of the libraries locally (OS X) as well as run them in Travis without having to worry. So much easier to iterate on it too (no global state getting polluted)

**Comment by maumueller** [comment]
Well, I guess I can lose my habit of working in VMs then. Great!

One question: How is the communication between the docker process and the "main process". Is the algorithm run going to create files which are then copied out of the container?

**Comment by erikbern** [comment]
> One question: How is the communication between the docker process and the "main process". Is the algorithm run going to create files which are then copied out of the container?

My plan is to pass the arguments to the child process on the command line. This is actually independent of whether we use Docker or not. But basically the parent process triggers `python run_algorithm.py --algorithm annoy --arg eucliean --arg 100` or something like that. I already refactored so that the child process reads the input data and stores the result.

Docker supports mounting directories on the parent's file system into the container, so there is nothing that has to be copied. We just have to mount `data`, `results`, and `ann_benchmarks` (the last one might not be strictly needed)

**Comment by maumueller** [comment]
> My plan is to pass the arguments to the child process on the command line. This is actually independent of whether we use Docker or not. But basically the parent process triggers python run_algorithm.py --algorithm annoy --arg eucliean --arg 100 or something like that. I already refactored so that the child process reads the input data and stores the result.
> 

Fair enough, each algorithm run really is a separate entity. So the only communication is spawning the (dockerized) process and waiting for it to terminate? Seems like a good separation of concerns. I hope that you will support a "time limit", that proved very useful to guard against enormously long index build times of the kNN graph variants. 

> Docker supports mounting directories on the parent's file system into the container, so there is nothing that has to be copied. We just have to mount data, results, and ann_benchmarks (the last one might not be strictly needed)

Great.

**Comment by erikbern** [comment]
> Fair enough, each algorithm run really is a separate entity. So the only communication is spawning the (dockerized) process and waiting for it to terminate? Seems like a good separation of concerns. I hope that you will support a "time limit", that proved very useful to guard against enormously long index build times of the kNN graph variants.

Yes, Docker lets you impose all kinds of limits (memory, cpu, time, whatnot) so that's definitely the plan. Probably 2 hours or something like that

**Comment by erikbern** [comment]
Last commit ends up breaking a whole bunch of stuff... I'm trying to get an end to end prototype working before I get everything 100% back working. In particular the result files end up having kind of broken paths... will revisit shortly.

**Comment by erikbern** [comment]
The cool thing is we're now running algorithms inside Docker containers

Main thing to be fixed is that since definitions are passed around by "reference", I can't call `name()` so I can't map runs to the result files properly. This breaks `plot.py` as well as storing what parameters already ran.

Other minor things:
* Resource limits on containers
* Catch nonzero exit statuses of child process
* Figure out how to stream stdout

**Comment by erikbern** [comment]
@maumueller is there a good reason why dicts are needed as parameters to the algorithm classes? i currently breaks kgraph and nmslib. Seems like it would be pretty easy to avoid (basically just destructure those dicts into a list of parameters) and would remove a bunch of complexity related to how algos.yml is parsed. Would also obviate the need for me to use some serialization format to pass data to the child process (can just send everything as command line arguments)

**Comment by maumueller** [comment]
@ale-f Can you elaborate on the dictionary choice? I'm sure I'm going to miss something here.

**Comment by ale-f** [comment]
`kgraph` took a dictionary before we started, so I didn't change that:

https://github.com/erikbern/ann-benchmarks/blob/711f6e88d71144c21241b8913b4662497b5a795e/ann_benchmarks/main.py#L554

The underlying constructors of `nmslib` take lists of parameter assignment strings (of the form `["param1=30", "param2=5"]`); the code used to construct these manually, which seemed unnecessarily complicated, so I replaced them with a dictionary that gets converted to this form when necessary.

https://github.com/erikbern/ann-benchmarks/blob/711f6e88d71144c21241b8913b4662497b5a795e/ann_benchmarks/main.py#L568

**Comment by ale-f** [comment]
(I suppose I should add that the algorithm definition file can't automatically make all combinations of these lists of parameter assignment strings, and I thought adding support for dictionaries, a more generally-useful data structure, and using those instead would be a good idea.)

**Comment by erikbern** [comment]
ok i decided to sidestep the question of dicts for now by using json as an intermediate representation – a bit hacky but should be fine

**Comment by erikbern** [comment]
* kgraph crashes in the constructor (segfault), not sure why. Tempted to remove it for now
* falconn times out on Travis, and I'm not able to compile it locally (think during compilation step the compiler runs out of memory). will try to fix
* dolphinnpy just super slow, i removed a few of the slow ones
*

**Comment by erikbern** [comment]
Going to fix the results stuff next, that's the only remaining part

**Comment by erikbern** [comment]
The result filenames are now computed directly from the algorithm definitions. Next (only remaining) step is that instead of listing the files on disk, just get all the algorithm definitions and compute their filenames

**Comment by erikbern** [comment]
@maumueller @ale-f take a look and see if this makes sense

it now runs all algos inside docker and plotting works. i ended up having to change a bit more than anticipated, but overall it's a net reduction of lines of code so i guess that's good

i don't think `createwebsite.py` works but it's probably not too hard to fix

**Comment by maumueller** [comment]
@erikbern I'll take a look at it Dec 13.

**Comment by erikbern** [comment]
ok, sounds good. will run some more tests until then

**Comment by maumueller** [comment]
I noticed the following when running the benchmarks (I consider all of them regressions from the current version at master):
- the current `install.sh` fails on `sh install.sh` (not providing a library), I think `sh install.sh` should  scan for `install/Dockerfile.(\w+)` and loop through all candidates
- I think install.sh should make sure that the local python installation provides everything necessary to run the benchmark. On a fresh amazon ec2 ubuntu that was `pip3 install h5py docker scipy ansicolors matplotlib`
- Without having certain algorithms installed invoking `python run.py` is a mess. I think it should check against the docker images to see which algorithms are available, before setting up the order in which each algorithm run occurs. The current version on master fails silently with a message "constructor not available" in such a case.
- unrelated to this pr: running with an invalid dataset (say glove-10-angular) results in an "urllib.error.HTTPError: HTTP Error 404: Not Found", I expected the provided argument is checked against the DATASETS dict we provide, and a message "dataset not available", maybe followed by a list of datasets that are available. (In general, a --list-datasets seems useful.)
- Should containers be cleaned up after a run? The list of containers is quite overwhelming even after a small benchmark. 

On the other hand:
- I'm currently working on including some GPU versions. Here, I would need to run nvidia-docker to get GPU bindings. Could that be somehow supported?

File: ann_benchmarks/algorithms/constructors.py [hunk 1]
```diff
@@ -1,59 +0,0 @@
...
-            available_constructors[symbol] = getattr(module, symbol)
-    except (ImportError, SyntaxError):
-        try:
-            t, v, tb = sys.exc_info()
-            traceback.print_exception(t, v, tb)
-        finally:
-            del tb
-        print('warning: module %s could not be loaded, some algorithm constructors will not be available' % name)
-        for symbol in symbols:
-            available_constructors[symbol] = None
```
Comment (maumueller) [inline ann_benchmarks/algorithms/constructors.py]: This was the code to fail silently when algorithms are not available. We need something like that again.

**Comment by erikbern** [comment]
see comments below (basically agree with all of them)

> the current install.sh fails on sh install.sh (not providing a library), I think sh install.sh should scan for install/Dockerfile.(\w+) and loop through all candidates

Yes, good point. Actually had the same thought earlier, forgot about it

> I think install.sh should make sure that the local python installation provides everything necessary to run the benchmark. On a fresh amazon ec2 ubuntu that was pip3 install h5py docker scipy ansicolors matplotlib

That's now handled by running `pip install -r requirements.txt` – since everything is just a Python package, you can run it in a virtual environment. I'll make sure to document it.

> Without having certain algorithms installed invoking python run.py is a mess. I think it should check against the docker images to see which algorithms are available, before setting up the order in which each algorithm run occurs. The current version on master fails silently with a message "constructor not available" in such a case.

That makes sense. I'll hit the Docker API first to list all images.

> unrelated to this pr: running with an invalid dataset (say glove-10-angular) results in an "urllib.error.HTTPError: HTTP Error 404: Not Found", I expected the provided argument is checked against the DATASETS dict we provide, and a message "dataset not available", maybe followed by a list of datasets that are available. (In general, a --list-datasets seems useful.)

Hm I thought I did that. Let me double check.

> Should containers be cleaned up after a run? The list of containers is quite overwhelming even after a small benchmark.

I kill the container after the benchmark is done, but you're right that I forgot to delete the container as well. Should do that, good point.

> Somehow unrelated:
> 
> I'm currently working on including some GPU versions. Here, I would need to run nvidia-docker to get GPU bindings. Could that be somehow supported?

Makes a lot of sense, and should be easy to support (I think). I'm guessing nvidia-docker provides the same API? The Docker Python package is really just a HTTP client so I think anything that's API compatible will work out of the box

**Comment by erikbern** [comment]
> Starting the docker container takes an enormous amount of time. Comparing running with `--local' and makes difference of about 1 minute for a single algorithm run. Is there something that we could improve in this respect?

What OS are you running? I run OS X with the newest version of Docker and it's super fast for me. Starting a container takes an unnoticeable amount of time and in fact running `time python3.6 run.py --docker-tag ann-benchmarks-annoy --max-n-algorithms 1 --dataset random-xs-20-angular` completes in about 15 seconds (most of which is spent inside Annoy itself).

File: ann_benchmarks/algorithms/annoy.py [hunk 2]
```diff
@@ -4,13 +4,13 @@
...
-        self._n_trees = n_trees
-        self._search_k = search_k
+        self._n_trees = int(n_trees)
+        self._search_k = int(search_k)
         self._metric = metric
-        self.name = 'Annoy(n_trees=%d, search_k=%d)' % (n_trees, search_k)
 
     def fit(self, X):
         self._annoy = annoy.AnnoyIndex(X.shape[1], metric=self._metric)
+        self._annoy.verbose(True)
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/algorithms/annoy.py]: sure can do

File: ann_benchmarks/algorithms/base.py [hunk 3]
```diff
@@ -3,37 +3,18 @@
...
-                    if unit == "B":
-                            val /= 1000.0
-                    if unit == "mB":
-                            val *= 1e3
-                    if unit == "gB":
-                            val *= 1e6
-                    return val
-        except:
-            print("Couldn't open status file, no index size available.")
-        return -1
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/algorithms/base.py]: Was planning to replace this with `container.stats()`:

https://docker-py.readthedocs.io/en/stable/containers.html
https://docs.docker.com/engine/reference/commandline/stats/#examples

I'll put something together. Forgot why I deleted this. I think at some point it was failing so I removed it thinking I'll fix it later

File: ann_benchmarks/algorithms/constructors.py (hunk 1 above)
Comment (erikbern) [reply ann_benchmarks/algorithms/constructors.py]: yeah – i'll list the containers when you invoke run.py and will warn about any missing containers

File: ann_benchmarks/algorithms/faiss.py [hunk 4]
```diff
@@ -8,13 +8,15 @@
...
 
     def fit(self, X):
-        X = X.astype(numpy.float32)
-        self._index = faiss.IndexLSH(len(X[0]), self._n_bits)
+        X = numpy.array(X)
+        if X.dtype != numpy.float32:
+            X = X.astype(numpy.float32)
+        f = X.shape[1]
+        self._index = faiss.IndexLSH(f, self._n_bits)
         self._index.train(X)
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/algorithms/faiss.py]: yeah good point. the h5py object is somewhat similar to the numpy array that for most algorithms you can pass it in (thanks to Python duck typing). but probably makes sense to just convert it to a numpy array in one place

File: ann_benchmarks/algorithms/itu.py [hunk 5]
```diff
@@ -19,7 +19,6 @@ def __init__(self, metric, alpha = None, beta = None, threshold = None, tau = No
         self._kappa2 = kappa2
         self._m1 = m1
         self._m2 = m2
-        self.name = ("ITUFilteringDouble(..., threshold = %f, ...)" % threshold)
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/algorithms/itu.py]: i can remove

File: ann_benchmarks/algorithms/nmslib.py [hunk 6]
```diff
@@ -29,12 +28,12 @@ def fit(self, X):
             # Aborted (core dumped)
             self._index_param.append('bucketSize=%d' % min(int(X.shape[0] * 0.0005), 1000))
                                         
-        self._index = nmslib.init(self._nmslib_metric, [], self._method_name, nmslib.DataType.DENSE_VECTOR, nmslib.DistType.FLOAT)
-    
+        self._index = nmslib.init(space=self._nmslib_metric,
+                                  method=self._method_name)
+
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/algorithms/nmslib.py]: i looked at the documentation of nmslib and those were the default parameters anyway. forgot why i removed them but i think it was something i had to do to get it running with the latest version at some point

**Comment by erikbern** [comment]
@maumueller did you look at the travis run btw? https://travis-ci.org/erikbern/ann-benchmarks/builds/314186734

pretty sweet to have CI for this going forward

File: ann_benchmarks/algorithms/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/algorithms/base.py]: actually the old code should be fine, i'll add it back. can't really use container.stats() from inside the container

**Comment by maumueller** [comment]
> > Starting the docker container takes an enormous amount of time. Comparing running with `--local' and makes difference of about 1 minute for a single algorithm run. Is there something that we could improve in this respect?

> What OS are you running? I run OS X with the newest version of Docker and it's super fast for me. Starting a container takes an unnoticeable amount of time and in fact running time python3.6 run.py --docker-tag ann-benchmarks-annoy --max-n-algorithms 1 --dataset random-xs-20-angular completes in about 15 seconds (most of which is spent inside Annoy itself).

My bad. The docker output is just not as responsive; the timings are indeed similar. 

```
$ time python3 run.py --docker-tag ann-benchmarks-annoy --max-n-algorithms 1 --dataset random-xs-20-angular --local

real	0m10.165s
user	0m11.760s
sys	0m3.024s

$ time python3 run.py --docker-tag ann-benchmarks-annoy --max-n-algorithms 1 --dataset random-xs-20-angular

real	0m10.788s
user	0m0.404s
sys	0m0.316s
```

**Comment by maumueller** [comment]
> @maumueller did you look at the travis run btw? https://travis-ci.org/erikbern/ann-benchmarks/builds/314186734

> pretty sweet to have continuous integration set up for this going forward

That's indeed great progress. (I do wonder if there would be a use in copying the generated plots.png back/somewhere as `plots-$LIBRARY-$commitid.png` to be able to locate problems.) We could even set it up in a way that other libraries set up a push hook on their master branch to invoke the travis build for their implementation.

I plan to include some basic unit tests for the metric generation over the holidays.

**Comment by maumueller** [comment]
>> Somehow unrelated:

>> I'm currently working on including some GPU versions. Here, I would need to run nvidia-docker to get GPU bindings. Could that be somehow supported?

> Makes a lot of sense, and should be easy to support (I think). I'm guessing nvidia-docker provides the same API? The Docker Python package is really just a HTTP client so I think anything that's API compatible will work out of the box

It seems to work fine. I can start and stop nvidia-docker containers from python-docker, so it should work.

**Comment by maumueller** [comment]
One unrelated fix: 

> To run experiments on SIFT, invoke `python run.py --dataset glove-100-angular`. See `python run.py --help` for more information on possible settings. Note that experiments can take a long time.

Change SIFT to GloVe. :-) 

Apart from this, no further suggestions. Great work.

**Comment by erikbern** [comment]
Travis can only run some super basic version though, so I don't know if you can really make sure it "works" – I guess you can have some lower threshold of expected accuracy maybe (I do this for Annoy's tests)

**Comment by maumueller** [comment]
Yeah, something like this would work. I mean, even a dataset consisting of one point and querying for that one point would make a difference to the current, kind of fragile setup.

**Comment by erikbern** [comment]
Right – Travis actually runs `plot.py` and it would also be fairly simple to just inspect the result files and make sure the kNN-precision is fairly high for the random-xs-20-angular dataset

**Comment by maumueller** [comment]
One small detail: It seems that that you removed the "name" attribute from an algorithm run. That is used in `createwebsite.py` to show which parameter values led to a certain data point (see below). Do you see another way of pretty printing the parameters? 
screen shot 2017-12-17 at 13 40 00

**Comment by erikbern** [comment]
if it's fine to show eg `FALCONN(16, 33, 33)` rather than with named arguments then it's very simple

can add back names but if the only use is the website plots then i'm not sure if it's worth the extra code

**Comment by maumueller** [comment]
Sure, that would be easy, but I think it should be annotated. 

I agree that we shouldn't add custom parameters, which are easily forgotten or outdated. We could inspect the constructor to get its parameter list and zip the parameters with their values. I suppose the constructor is accessible from the information in the `algos.yaml` file?

**Comment by erikbern** [comment]
the constructor is not accessible since importing respective algorithm will in turn import respective library – those are not available outside the docker container. so it's not that easy.

the only thing right now that i can think of is if each algorithm writes the name into the result hdf5 file

let me know if you feel strongly about this, i'd rather leave it outside the scope of this PR so that we can merge this

**Comment by maumueller** [comment]
> the only thing right now that i can think of is if each algorithm writes the name into the result hdf5 file

That's exactly what the current version in master does and I see nothing wrong about it. (If we cannot access the constructor from our plotting tools to do it for us in a cleverer way.)

> let me know if you feel strongly about this, i'd rather leave it outside the scope of this PR so that we can merge this. happy to think about solutions later.

Honestly, the scope of this PR is then to _remove_ this feature. The safe way would be to revert the changes to `algo.name` and think about solutions later.

**Comment by erikbern** [comment]
ok let me revert the `.name()` stuff since you feel strongly about it

**Comment by erikbern** [comment]
btw running this on a c4.2xlarge for a few days and i think the 8GB limit is too strict – going to set the memory limit dynamically based on the host memory available instead

**Comment by maumueller** [comment]
Thanks for leaving `.name()` in. When are you going to merge this into master? I want to get `createwebsite.py` up and running  as well, but I think it should be a pull request against master.

**Comment by erikbern** [comment]
i can merge this in a few days. afaik the only things that are missing are

1. `name()` thing
2. finish running benchmarks and make sure they look good

**Comment by erikbern** [comment]
@maumueller this OK to merge? I addressed your concerns

**Comment by erikbern** [comment]
I'm struggling to run kgraph – it crashes sporadically that I'm not sure about. Don't think it has anything to do with the Dockerization though

**Comment by erikbern** [comment]
image

Some notes

* FALCONN is broken (again?)
* didn't get kgraph to run, it keeps crashing
* not sure what's up with panns? should prob just drop it
* annoy disappointingly slow, not sure why

**Comment by erikbern** [comment]
I'll merge this in a couple of days unless anyone objects @maumueller @ale-f

**Comment by erikbern** [comment]
This PR has been going on for too long, and I'm 100% convinced Dockerization is the way to go, so I merged it and we'll deal with random fallout as follow up fixes

**Pull request merged**